sdk/include/game-contract-sdk/game_base.hpp

```diff
@@ -112,149 +120,163 @@ class game: public eosio::contract {
+    virtual void on_new_game(uint64_t ses_id) = 0; // must be overridden
+    virtual void on_action(uint64_t ses_id, uint16_t type, std::vector<uint32_t> params) = 0; // must be overridden
+    virtual void on_random(uint64_t ses_id, checksum256 rand) = 0; // must be overridden
+    virtual void on_finish(uint64_t ses_id) { /* do nothing by default */ } // optional
 
-public:
     /* getters */
-    global_row get_global() {
-        return global.get_or_default();
+    const global_row& get_global() const {
+        return global;
     }
 
-    const session_row& get_session(uint64_t req_id) const {
-        return sessions.get(req_id);
+    const session_row& get_session(uint64_t ses_id) const {
+        return sessions.get(ses_id, "session with this ses_id not found");
     }
 
-    std::optional<uint32_t> get_param_value(uint64_t req_id, uint16_t param_type) const {
-        const auto& session = sessions.get(req_id);
+    std::optional<uint32_t> get_param_value(uint64_t ses_id, uint16_t param_type) const {
+        const auto& session = sessions.get(ses_id);
         const auto itr = std::find_if(session.params.begin(), session.params.end(),
         [&](const auto& item){
             return item.first == param_type;
         });
         return itr == session.params.end() ? std::nullopt : std::optional<uint32_t> { itr->second };
     }
 
-public:
+protected:
     /* utility helpers */
-    uint128_t rand_u128(const checksum256& rand) {
+    uint128_t rand_u128(const checksum256& rand) const {
         const auto& arr = rand.get_array();
-        return arr[0] / 2 + arr[0] / 2;
+        // use % operation to save original distribution
+        const uint128_t left = arr[0] % UINT64_MAX;
+        const uint128_t right = arr[1] % UINT64_MAX;
+
+        // just concat parts
+        // it's not fair way(don't save original distribution), but more simpler
+        return (left << 64) & right;
     }
 
-    uint64_t rand_u64(const checksum256& rand) {
+    uint64_t rand_u64(const checksum256& rand) const {
         auto u128 = rand_u128(rand);
-        return ((uint64_t)(u128 >> 64) / 2) + ((uint64_t)u128 / 2);
+        return u128 % UINT64_MAX;
     }
 
-public:
+protected:
     /* game session state changers */
-    void require_action(uint64_t req_id, uint8_t action_type) {
-        const auto session_itr = sessions.require_find(req_id, "session with this req_id not found");
-        eosio::check(session_itr->state == static_cast<uint8_t>(state::req_start) ||
-                     session_itr->state == static_cast<uint8_t>(state::req_signidice_part_2),
-        "state should be 'req_start' or 'req_signidice_part_2'");
-
-        sessions.modify(session_itr, get_self(), [&](auto& obj){
-            obj.state = static_cast<uint8_t>(state::req_action);
+    void require_action(uint64_t ses_id, uint8_t action_type, bool need_random = false) {
+        const auto& session = get_session(ses_id);
+
+        check_current_session(session);
```

Comment:
Instead of doing `check_current_session` here and there can we remove the `uint64_t ses_id` parameter altogether?
A child class will deal with the `current_session` by definition.

Reply:
It's good idea to implicitly get `ses_id` from `current_session`, so I've done it in commit below.